lib/DoctrineCodingStandard/Sniffs/Classes/ExceptionClassNamingSniff.php

```diff
@@ -0,0 +1,99 @@
+<?php
+
+declare(strict_types=1);
+
+namespace DoctrineCodingStandard\Sniffs\Classes;
+
+use DoctrineCodingStandard\Helpers\UseStatementHelper;
+use PHP_CodeSniffer\Files\File;
+use PHP_CodeSniffer\Sniffs\Sniff;
+use const T_ABSTRACT;
+use const T_CLASS;
+use const T_FINAL;
+use function count;
+use function preg_grep;
+use function preg_match;
+
+final class ExceptionClassNamingSniff implements Sniff
+{
+    private const CODE_NOT_AN_EXCEPTION_CLASS = 'NotAnExceptionClass';
+
+    /**
+     * {@inheritdoc}
+     */
+    public function register() : array
+    {
+        return [T_CLASS];
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function process(File $phpcsFile, $stackPtr) : void
+    {
+        $isAbstract              = $phpcsFile->findFirstOnLine([T_ABSTRACT], $stackPtr) !== false;
+        $isFinal                 = $phpcsFile->findFirstOnLine([T_FINAL], $stackPtr) !== false;
+        $isExtendingException    = $this->isExtendingException($phpcsFile, $stackPtr);
+        $isImplementingException = $this->isImplementingException($phpcsFile, $stackPtr);
+        $hasExceptionName        = $this->hasExceptionSuffix((string) $phpcsFile->getDeclarationName($stackPtr));
+        $hasValidClassName       = ($isAbstract && $hasExceptionName) ||
+            (! $isAbstract && ! $hasExceptionName && $isFinal);
+        $isValidException        = $hasValidClassName && ($isExtendingException || $isImplementingException);
+        $isNoException           = ! $hasExceptionName && ! $isExtendingException && ! $isImplementingException;
```

Comment:
Please try to turn these combinations into multiple early-exit `if`s, this is too complex.